examples/examples/jsonrpsee_service.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::error::Error as StdError;
+use std::net::SocketAddr;
+use std::sync::atomic::AtomicU32;
+use std::sync::Arc;
+
+use futures::FutureExt;
+use jsonrpsee::core::{async_trait, client::ClientT};
+use jsonrpsee::http_client::HttpClientBuilder;
+use jsonrpsee::proc_macros::rpc;
+use jsonrpsee::rpc_params;
+use jsonrpsee::server::{http, ConnectionGuard, ServerHandle, StopHandle, TowerService};
+use jsonrpsee::types::ErrorObjectOwned;
+
+use hyper::server::conn::AddrStream;
+use tower::Service;
+use tracing_subscriber::util::SubscriberInitExt;
+
+#[rpc(server)]
+pub trait Rpc {
+	#[method(name = "say_hello")]
+	async fn say_hello(&self) -> Result<String, ErrorObjectOwned>;
+}
+
+#[async_trait]
+impl RpcServer for () {
+	async fn say_hello(&self) -> Result<String, ErrorObjectOwned> {
+		Ok("lo".to_string())
+	}
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+	let filter = tracing_subscriber::EnvFilter::try_from_default_env()?
+		.add_directive("jsonrpsee[method_call{name = \"say_hello\"}]=trace".parse()?);
+	tracing_subscriber::FmtSubscriber::builder().with_env_filter(filter).finish().try_init()?;
+
+	let handle = run_server();
+	tokio::spawn(handle.stopped());
+
+	let client = HttpClientBuilder::default().build("http://127.0.0.1:9944").unwrap();
+
+	let x: String = client.request("say_hello", rpc_params!()).await.unwrap();
+	tracing::info!("response: {x}");
+
+	Ok(())
+}
+
+fn run_server() -> ServerHandle {
```

Comment:
Aah ok, so this is the "minimal" example of using `jsonrpsee` as a tower service; this is interesting!

We are exposing a bunch of things I guess in order to make this possible; I wonder whether it's possible to hide any of this stuff without reducing the flexibility of this approach in general! Eg maybe a builder struct or something to make constructing the `TowerService` easier (or a `TowerServiceBuilder` that can be turned into a new `TowerService` on each connection or something). 

Partly I'm just wondering whether we _want_ to expose some of the things that we have in jsonrpsee, or whether it's ok that we exposed the extra bits? And partly just thinking how to make this super easy in the "simple" case but allowing for flexibility when needed!

Reply:
I realized that it won't play that nicely with our http middleware such that one would need to use the low-level API in for hyper that.

Thus, I believe that this would require less code https://github.com/hyperium/hyper/blob/master/examples/service_struct_impl.rs

Reply:
Fair point, I made everything public because it was an easy way to extract all settings from the ServerBuilder to avoid having 10 parameters for the user to inject.

Lemme a think about it

Reply:
I have tried to implement something cleaner and I removed the jsonrpsee::to_service from the low-level api example